### PR TITLE
vendor: add go-mod support

### LINKED ISSUE #5367: proposal: explore use of go modules for dependencies in Cilium
Now that go 1.11 has been released ( https://golang.org/doc/go1.11 ) we can explore whether using go modules is feasible for Cilium: https://golang.org/doc/go1.11#modules

### LINKED ISSUE #9169: Go modules

### PR DESCRIPTION
**Pull request by aanm**
```
    As a consequence of doing updates for all dependencies some changes had
    to be performed:
    - pkg/aws/ec2/ec2.go add context to function calls
    - pkg/kvstore change import path from go.etcd.io/etcd to github.com/coreos/etcd
    - add tools package to import non-go dependencies into vendor.
    - add documentation to update dependencies
```

Signed-off-by: André Martins <andre@cilium.io>

Fixes: https://github.com/cilium/cilium/issues/5367
Fixes: #9169

```release-note
add support for go modules
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8719)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-missed-k8s

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by ianvernon** [comment]
@aanm when can this be merged?

**Comment by aanm** [comment]
never-tell-me-the-odds

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
never-tell-me-the-odds

**Comment by aanm** [comment]
never-tell-me-the-odds

File: Makefile [hunk 1]
```diff
@@ -6,9 +6,9 @@ ifdef LIBNETWORK_PLUGIN
 SUBDIRS_CILIUM_CONTAINER += plugins/cilium-docker
 endif
 SUBDIRS := $(SUBDIRS_CILIUM_CONTAINER) operator plugins tools
-GOFILES_EVAL := $(subst _$(ROOT_DIR)/,,$(shell $(CGO_DISABLED) $(GO) list ./... | grep -v -e /vendor/ -e /contrib/))
+GOFILES_EVAL := $(subst _$(ROOT_DIR)/,,$(shell $(CGO_DISABLED) $(GOLIST) $(GO) list -e ./...))
 GOFILES ?= $(GOFILES_EVAL)
-TESTPKGS_EVAL := $(subst github.com/cilium/cilium/,,$(shell $(CGO_DISABLED) $(GO) list ./... | grep -v '/api/v1\|/vendor\|/contrib' | grep -v -P 'test(?!/helpers/logutils)'))
+TESTPKGS_EVAL := $(subst github.com/cilium/cilium/,,$(shell $(CGO_DISABLED) $(GOLIST) $(GO) list -e ./... | grep -v -P 'test(?!/helpers/logutils)'))
```
Comment (joestringer) [inline Makefile]: Do you know why we excluded `api/` before? Does it make a meaningful difference to test run time?

File: contrib/go-mod/update-vendor.sh [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+ensure_require_replace_directives_for_all_dependencies
+# group replace directives
+group_replace_directives
+
+go mod tidy >>"${LOG_FILE}" 2>&1
+
+#add_generated_comments "
+#// This is a generated file. Do not edit directly.
+#// Run hack/pin-dependency.sh to change pinned dependency versions.
+#// Run hack/update-vendor.sh to update go.mod files and the vendor directory.
```
Comment (joestringer) [inline contrib/go-mod/update-vendor.sh]: I think these are now under `contrib/`, not `hack/`?

File: hack/82410.patch [hunk 3]
```diff
@@ -0,0 +1,3 @@
+diff --git a/vendor/k8s.io/code-generator/generate-groups.sh b/vendor/k8s.io/code-generator/generate-groups.sh
+old mode 100644
+new mode 100755
```
Comment (joestringer) [inline hack/82410.patch:3]: Why is this necessary? Can we upstream it?

**Review by joestringer — APPROVED** [review]
Completely trivial nits below. And how terrible is GitHub at handling commits like this.. LGTM. 

For any other reviewers, here's how I looked at the meaningful changes (assuming that `vendor/` was correctly updated):

```
$ git show upstream/pr/add-go-mod-support -- :^vendor
```

**Comment by joestringer** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

File: hack/82410.patch (hunk 3 above)
Comment (aanm) [reply hack/82410.patch:3]: good point, I didn't even realize this was a bug... https://github.com/golang/go/issues/34965

File: Makefile (hunk 1 above)
Comment (aanm) [reply Makefile]: odd this might have been removed by accident, re-added.

**Comment by aanm** [comment]
test-me-please
